<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="wavClusteR : An R/BioConductor package for PAR-CLIP data analysis">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>wavClusteR</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/FedericoComoglio/wavClusteR">View on GitHub</a>

          <h1 id="project_title">wavClusteR</h1>
          <h2 id="project_tagline">An R/BioConductor package for PAR-CLIP data analysis</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/FedericoComoglio/wavClusteR/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/FedericoComoglio/wavClusteR/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h2>
<a id="background" class="anchor" href="#background" aria-hidden="true"><span class="octicon octicon-link"></span></a>Background</h2>

<h2>
<a id="what-is-par-clip" class="anchor" href="#what-is-par-clip" aria-hidden="true"><span class="octicon octicon-link"></span></a>What is PAR-CLIP?</h2>

<p>The Photoactivatable-Ribonucleoside-Enhanced Crosslinking and Immunoprecipitation (<strong>PAR-CLIP</strong>) is a recently developed biochemical method used to map RNA-binding proteins (RBP) - RNA interactions transcriptome-wide and at high resolution. </p>

<p>In the standard PAR-CLIP protocol, cells are first cultured in the presence of a photoreactive ribonucleoside analogs, such as 4-thiouridine (4-SU) and 6-thioguanosine (6-SG), which is incorporated <em>in vivo</em> into nascent RNA transcripts. Irradiation with UV light (365 nm) is then used to covalently cross link the photoreactive nucleoside-labeled cellular RNAs to interacting RBPs. The cross linked RNA-RBP complexes are then isolated by immunoprecipitation of the RBP of interest and the coimmunoprecipitated RNA released by protein degradation. The resulting RNA is converted into a cDNA library and the identity of the RNA molecules is determine by next-generation sequencing (NGS). </p>

<p>Over the past five years, the PAR-CLIP protocol has been widely used to identify the binding sites of several known RBPs and microRNA-containing ribonucleoprotein complexes. However, PAR-CLIP data analysis is not straightforward and dedicated methods are required to control false discovery rates.</p>

<h2>
<a id="introduction" class="anchor" href="#introduction" aria-hidden="true"><span class="octicon octicon-link"></span></a>Introduction</h2>

<p>wavClusteR is an R/BioConductor package for PAR-CLIP data analysis can be used to:</p>

<ul>
<li>Infer PAR-CLIP induced transitions and discriminate them from sequencing errors, SNPs, contaminants and additional non-experimental causes, using a non-parametric mixture model (Sievers et al, 2012).</li>
<li>Resolve protein binding sites (clusters) at high resolution.</li>
<li>Rank clusters for downstream analyses and validation based on a solid statistical framework.</li>
<li>Integrate total RNA-Seq data, if available, to estimate false discovery rates.</li>
<li>Post-processing of clusters, including annotation of protein binding sites, data export in common UCSC formats and motif search analysis. </li>
</ul>

<p>The package <em>supports multicore computing</em>. While wavClusteR was designed for PAR-CLIP data analysis, it can be adapted to the analysis of other NGS data from experimental procedures exploting nucleotide substitutions for site detection (e.g. BisSeq). </p>

<h2>
<a id="version" class="anchor" href="#version" aria-hidden="true"><span class="octicon octicon-link"></span></a>Version</h2>

<p>Current stable version is 2.0</p>

<h2>
<a id="download-and-installation" class="anchor" href="#download-and-installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Download and Installation</h2>

<p>To install wavClusteR, start R (&gt;3.1.0) and enter:</p>

<div class="highlight highlight-R"><pre>source(<span class="pl-s1"><span class="pl-pds">"</span>http://bioconductor.org/biocLite.R<span class="pl-pds">"</span></span>)
biocLite(<span class="pl-s1"><span class="pl-pds">"</span>wavClusteR<span class="pl-pds">"</span></span>)</pre></div>

<h2>
<a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h2>

<p>The package provides a vignette detailing all steps required to analyze your PAR-CLIP data. To view this document for the version of wavClusteR installed in your system, start R and enter:</p>

<div class="highlight highlight-R"><pre>browseVignettes(<span class="pl-s1"><span class="pl-pds">"</span>wavClusteR<span class="pl-pds">"</span></span>)</pre></div>

<h2>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Citation</h2>

<p>If you are using wavClusteR for your analysis, please cite:</p>

<ul>
<li>Comoglio, F., Sievers, C., and Paro, R. (2015). Sensitive and highly resolved identification of RNA-protein interaction sites in PAR-CLIP data. BMC Bioinformatics, in the press.</li>
<li>Sievers, C., Schlumpf, T., Sawarkar, R., Comoglio, F., and Paro, R. (2012). Mixture models and wavelet transforms reveal high confidence RNA-protein interaction sites in MOV10 PAR-CLIP data. Nucleic Acids Res. 40(20), e160. <a href="http://www.ncbi.nlm.nih.gov/pubmed/22844102">doi 10.1093/nar/gks697</a>
</li>
</ul>

<h2>
<a id="contact" class="anchor" href="#contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Contact</h2>

<p>Should you</p>

<ul>
<li>experience any issue using wavClusteR</li>
<li>need support for your analysis</li>
<li>have specific operating system requirements</li>
</ul>

<p>please do not hesitate to contact me @ <a href="mailto:federico.comoglio@bsse.ethz.ch">federico.comoglio@bsse.ethz.ch</a></p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">wavClusteR maintained by <a href="https://github.com/FedericoComoglio">FedericoComoglio</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
